File: server/apollo/src/query-summary.ts
import { TransactionStatement } from '@tally/lib/statement/transaction';
import { combineSummaryStatements } from '@tally/lib/statement/summary';
import { toGqlStatement, toGqlSummaryStatement } from './to-gql';
import { loadBudget } from '@tally/lib/data/loader';
import {
  GqlSummaryData,
  QuerySummaryArgs,
} from './types';


export async function buildSummaryData(_: any, args: QuerySummaryArgs): Promise<GqlSummaryData> {
  const startTimeMs: number = Date.now();
  const payload = await loadBudget();
  const summaryName = args.accountType.startsWith('/') ? args.accountType :
    (args.owner + ' ' + (args.accountType === args.owner ? 'SUMMARY' : args.accountType));
  const monthSummaries = payload.summaries.get2(args.owner, summaryName);
  if (!monthSummaries) {
    throw new Error(`Summary ${args.accountType} for ${args.owner} not found.`);
  }
  const summaryStatements = [...monthSummaries.values()].filter(
    (stmt) => stmt.month.isBetween(args.startMonth, args.endMonth)
  );
  if (!summaryStatements.length) {
    throw new Error(
      `Summary ${args.accountType} for ${args.owner} for months [${args.startMonth}, ${args.endMonth}] not found.`
    );
  }
  const summary =
    summaryStatements.length === 1
      ? summaryStatements[0]
      : combineSummaryStatements(summaryStatements);
  const result = {
    statements: summary.statements
      .sort((a, b) => (a.account.name < b.account.name ? -1 : 1))
      .map((stmt) => toGqlStatement(stmt as TransactionStatement)),
    total: toGqlSummaryStatement(summary)
  };
  console.log(
    `gql "${summaryName}" summary data in ${Date.now() - startTimeMs}ms for [${args.startMonth}, ${args.endMonth}]`
  );
  return result;
}

File: server/kt/src/main/kotlin/schema/QuerySummary.kt
package com.cherba29.tally.schema

import com.cherba29.tally.core.Month
import com.cherba29.tally.data.DataPayload
import com.cherba29.tally.statement.TransactionStatement
import com.cherba29.tally.statement.combineSummaryStatements
import io.github.oshai.kotlinlogging.KotlinLogging
import kotlin.time.Clock

fun buildSummaryData(payload: DataPayload, owner: String, accountType:String, startMonth: Month, endMonth: Month): GqlSummaryData {
  // TODO: replace with measureTime.
  val startTimeMs: Long = Clock.System.now().toEpochMilliseconds()
  val summaryName = if (accountType.startsWith("/")) accountType else owner + " " + (if (accountType === owner) "SUMMARY" else accountType)
  // TODO: replace with indexing operator.
  val monthSummaries = payload.summaries.get2(owner, summaryName)
  if (monthSummaries == null) {
    throw NotFoundException("Summary $accountType for $owner not found.")
  }
  val summaryStatements = monthSummaries.values.filter { stmt -> stmt.month in startMonth..endMonth }
  if (summaryStatements.isEmpty()) {
    throw NotFoundException(
        "Summary $accountType for $owner for months [$startMonth, $endMonth] not found."
        )
  }
  val summary = if (summaryStatements.size == 1) summaryStatements.first() else combineSummaryStatements(summaryStatements)
  val result = GqlSummaryData(
    statements = summary.statements.sortedWith { a, b -> if (a.account.name < b.account.name) -1 else 1 }.map { stmt -> (stmt as TransactionStatement).toGql()},
    total = summary.toGql()
  )
  logger.info {
    "gql '${summaryName}' summary data in ${Clock.System.now().toEpochMilliseconds() - startTimeMs}ms for [$startMonth, $endMonth]"
  }
  return result
}

private val logger = KotlinLogging.logger {}